### DISCUSSION TITLE
Inserting JSON child without copying

### DISCUSSION THREAD

**Discussion by MGraefe** [nlohmann/json#2517] [Q&A] [answered]
Hi, I have the following code:

```cpp
std::string serializeChild(const nlohmann::json& child)
{
  // insert child into some parent structure with additional information before serializing
  nlohmann::json parent;
  parent["version"] = 42;
  parent["child"] = child; // copy occuring here :(
  return parent.dump(2);
}
```

Is it possible to perform this operation without _copying_ `child` into the parent object? So just insert a pointer/reference to the child into the parent, serialize and that's it? I had a look at `json_ref` but I think that's just for optimization while initializer lists are created.

**Comment by nlohmann** [comment, marked as answer]
No, this is not possible.

**Reply by epvbergen** [reply]
But wouldn't it be wonderful if it would? 

I coincidentally wrote a similar entry: https://github.com/nlohmann/json/discussions/2511

**Reply by nlohmann** [reply]
Beside runtime problems (dangling references), this approach would also would not work for implicit conversions. For instance, you can now write code like

```cpp
json j;
std::vector<int> v = {1,2,3};

j["vector"] = v;
```

To my knowledge, there is no way to support this with a reference as you would need to know the type when defining the value type of `json`.

**Reply by epvbergen** [reply]
The reference would be to a known type. And the risk and benefits of weak reference are no different from those in any other datastructure. The selling point of C++ is that it's the user making those decisions, not the language or its frameworks. 

I also hugely appreciate that basic_json::get_ref exists, which allows me to skip a copy when I merely need to look at a std::string in cases where they can get large, and don't need allocation, managing ownership, and all that jazz, when I know that the parsed object will live long and prosper way beyond my own lifetime.

**Reply by nlohmann** [reply]
I'll think about it.

**Reply by nlohmann** [reply]
> The reference would be to a known type.

What do you mean?

**Reply by MGraefe** [reply]
Since I am mainly interested in serialization, maybe instead of a reference type we could set custom serialization functions on the json object? Something like this:

```cpp
using nlohmann::json;
std::string serializeChild(const json& child)
{
  json parent;
  parent["version"] = 42;
  parent["child"].serializeFunc = [&](const json& obj) { return child.dump(2); };
  return parent.dump(2);
}
```

The caller is clearly responsible for what happens in the function when the json object is serialized, his fault if that's messed up / has dangling references.

Of course the json object would need a function pointer / std::function object, not sure if you're willing to do that. But I think it would be pretty neat for getting rid of this copy-before-serializing and could also be used for other stuff.
